File: experimental/sso-with-skills/csharp_dotnetcore/SkillBot/Dialogs/SsoSkillSignInDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;

namespace Microsoft.BotBuilderSamples.SkillBot.Dialogs
{
    public class SsoSkillSignInDialog : ComponentDialog
    {
        public SsoSkillSignInDialog(string connectionName)
            : base(nameof(SsoSkillSignInDialog))
        {
            AddDialog(new OAuthPrompt(nameof(OAuthPrompt), new OAuthPromptSettings
            {
                ConnectionName = connectionName,
                Text = "Sign in to the Skill using Azure AD",
                Title = "Sign In"
            }));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { SignInStepAsync, DisplayTokenAsync }));
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> SignInStepAsync(WaterfallStepContext context, CancellationToken cancellationToken)
        {
            // This prompt won't show if the user is signed in to the root using SSO.
            return await context.BeginDialogAsync(nameof(OAuthPrompt), null, cancellationToken);
        }

        private async Task<DialogTurnResult> DisplayTokenAsync(WaterfallStepContext context, CancellationToken cancellationToken)
        {
            if (!(context.Result is TokenResponse result))
            {
                await context.Context.SendActivityAsync("No token was provided for the skill.", cancellationToken: cancellationToken);
            }
            else
            {
                await context.Context.SendActivityAsync($"Here is your token for the skill: {result.Token}", cancellationToken: cancellationToken);
            }

            return await context.EndDialogAsync(null, cancellationToken);
        }
    }
}

File: samples/javascript_nodejs/82.skills-sso-cloudadapter/skillBot/dialogs/ssoSkillSignInDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ComponentDialog, OAuthPrompt, WaterfallDialog } = require('botbuilder-dialogs');

const OAUTH_PROMPT = 'OAuthPrompt';
const WATERFALL_DIALOG = 'WaterfallDialog';
const SSO_SIGNIN_DIALOG = 'SsoSkillSignInDialog';

class SsoSkillSignInDialog extends ComponentDialog {
    /**
     * @param {string} connectionName
     */
    constructor(connectionName) {
        super(SSO_SIGNIN_DIALOG);

        this.addDialog(new OAuthPrompt(OAUTH_PROMPT, {
            connectionName: connectionName,
            text: 'Sign in to the Skill using Azure AD',
            title: 'Sign In'
        }));

        this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
            this.signInStep.bind(this),
            this.displayToken.bind(this)
        ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async signInStep(stepContext) {
    // This prompt won't show if the user is signed in to the root using SSO.
        return stepContext.beginDialog(OAUTH_PROMPT);
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async displayToken(stepContext) {
        const { result } = stepContext;
        if (!result || !result.token) {
            await stepContext.context.sendActivity('No token was provided for the skill.');
        } else {
            await stepContext.context.sendActivity(`Here is your token for the skill: ${ result.token }`);
        }

        return stepContext.endDialog();
    }
}

module.exports.SsoSkillSignInDialog = SsoSkillSignInDialog;
